### ISSUE TITLE
`and` operator does not correctly handle inheritance

### ISSUE THREAD

**Issue by iwahbe** [python/mypy#12030]
**Bug Report**

mypy is providing false positives when subclassing in `and` statements. 

**To Reproduce**

Create a file repro.py:
```python
#!/usr/bin/env python3

class Foo:
    def __init__(self):
        pass
class Bar(Foo):
    def __init__(self):
        super().__init__()

for foo in [Foo(), None]:
    for bar in [Bar(), None]:
        reveal_type(foo and bar)
        print(foo and bar)
```

Run mypy:
```sh
$ python3 -m mypy repro.py
repro.py:12: note: Revealed type is "Union[mypy_bug.Foo, None]"
```

**Expected Behavior**

Revealed type should be `"Union[None, repro.Bar]"`.

**Actual Behavior**

Revealed type is `"Union[None, repro.Foo]"`.

**Your Environment**

- Mypy version used: mypy 0.931
- Mypy command-line flags: None (see above)
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: Python 3.9.9
- Operating system and version: MacOS 12.1 (21C52)

**Comment by A5rocks** [comment]
I believe this is intended behavior: when mypy joins together types, it does so into a single type: `repro.Bar` + `repro.Foo` = `repro.Foo` and not `Union[repro.Bar, repro.Foo]`. The latter would give you (somewhat) what you expected, as in `Union[None, repro.Bar, repro.Foo]`. (just `Optional[repro.Bar]` would not be OK, as `repro.Foo` does not necessarily implement everything `repro.Bar` does just because of the subclassing relationship)

**Comment by erictraut** [comment]
Are you sure this is intended behavior? This doesn't look like a join issue. It looks like a legitimate bug to me.

If the first operand is truthy (such as `Foo`), the `and` operator always returns the second operand. And if the first operand is falsy (such as `None`) it always returns the first operand. Since the first operand in this case has the type `None | Foo`, mypy should be able to determine that the final type will be either `None` or the type of the second operand (`Bar | None`). When you combine `None` and `Bar | None`, it should produce the answer `Bar | None`. (This is the answer that pyright produces.)

**Comment by A5rocks** [comment]
Oh wait sorry, I confused myself. Geh. You're very right.

**Comment by iwahbe** [comment]
@A5rocks You are right. We can confirm that this is unsafe by implementing 
```python
__bool__(self):
    return False
```
for Foo. We then see
```sh
<__main__.Foo object at 0x1056adee0>
<__main__.Foo object at 0x1056adee0>
None
None
```
It would be great if mypy was able to check if Foo (or subclasses) implement `__bool__`, but that sound like a feature request (if it's feasible), and not a bug report.
